### PR TITLE
Consume STOMP message and publish to RabbitMQ

### PR DESCRIPTION
**Pull request by datanel**
`internal` directory contains webstomp and rabbitmq client. `internal` means that lib is not exposed and cannot be used outside.
if we want to expose the lib we could rename `internal` to `pkg`

## TODO (may be in another pull requests)

- [x] better encapsulate the code about Stomp
- [ ] handle websocket/stomp reconnection
- [ ] It seems that interruption (ctrl +C, SIGKILL/SIGINT)  is not well handled. I have `panic: runtime error: invalid memory address or nil pointer dereference`

### CODE REVIEW

File: cmd/stomptherabbit/main.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+
+		<-c
+		close(done)
+	}()
+
+	var conn *stomp.Conn
+	go func() {
+	Loop:
```
Comment (pascal-vs) [inline cmd/stomptherabbit/main.go:54]: I think it works even without this label.
The `continue` will go directly to the next step of this infinite loop

File: cmd/stomptherabbit/main.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+				if err != nil {
+					log.Printf("failed to aknowledge message, err: %v\n", err)
+				}
+			}
+		}
+	}()
+	fmt.Println(c.ToString())
+	fmt.Println("Waiting for messages...")
+	<-done
+	conn.Disconnect()
```
Comment (pascal-vs) [inline cmd/stomptherabbit/main.go:90]: maybe a log when a normal exit happens?

File: cmd/stomptherabbit/config.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	}
+}
+
+func (c *Config) ToString() string {
+	var b strings.Builder
+
+	fmt.Fprintln(&b, "Webstomp configuration")
+	fmt.Fprintln(&b, "----------------------")
+	fmt.Fprintln(&b, "Webstomp target is\t\t", c.Webstomp.Target)
+	fmt.Fprintln(&b, "Webstomp loging is\t\t", c.Webstomp.Login)
```
[the remark this answers is not in the document]
Comment (datanel) [reply cmd/stomptherabbit/config.go:35]: this function is a special dedication for you @pascal-vs

File: cmd/stomptherabbit/config.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	viper.AutomaticEnv()
+	// viper.BindEnv("WEBSTOMP")
+	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
+
+	viper.SetDefault("webstomp.sendTimeout", "0")
+	viper.SetDefault("webstomp.recvTimeout", "0")
+	viper.SetDefault("rabbitmq.url", "amqp://guest:guest@localhost:5672//")
+	viper.SetDefault("rabbitmq.contentType", "application/json")
+
+	// I need to that to be able to unmarshal from env vars
```
[the remark this answers is not in the document]
Comment (datanel) [reply cmd/stomptherabbit/config.go]: fixed

File: cmd/stomptherabbit/config.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	viper.BindEnv("webstomp.login")
+	viper.BindEnv("webstomp.passcode")
+	viper.BindEnv("webstomp.protocol")
+	viper.BindEnv("webstomp.destination")
+	viper.BindEnv("rabbimq.url")
+	viper.BindEnv("rabbitmq.exchange.name")
+}
+
+func config() (*Config, error) {
+	if err := viper.ReadInConfig(); err != nil {
```
[the remark this answers is not in the document]
Comment (datanel) [reply cmd/stomptherabbit/config.go:75]: yes, fixed

File: cmd/stomptherabbit/main.go (hunk 1 above)
Comment (datanel) [reply cmd/stomptherabbit/main.go]: you are right fixed

File: cmd/stomptherabbit/main.go (hunk 2 above)
Comment (datanel) [reply cmd/stomptherabbit/main.go:90]: yes added

File: config.yml.example [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  passcode: "your_passcode"
+  destination: "/topic/a_topic"
+  protocol: "v12.stomp"
+  sendTimeout: 20000
+  recvTimeout: 20000
+rabbitmq:
+  url: amqp://guest:guest@localhost:5672//
+  contentType: application/json
+  exchange:
+    name: your_exhange_name
```
[the remark this answers is not in the document]
Comment (datanel) [reply config.yml.example]: I don't think it's a problem if there is no space in the string. But I can add quotes everywhere

File: internal/webstomp/webstomp.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	}
+	origin, err := u.Parse("/")
+	if err != nil {
+		return nil, err
+	}
+	origin.Scheme = "https"
+	c, err := websocket.Dial(u.String(), opts.Protocol, origin.String())
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to open a new client connection to a the WebSocket %s, err: %v", target, err)
```
Comment (woshilapin) [inline internal/webstomp/webstomp.go:34]:
```suggestion
		return nil, fmt.Errorf("failed to open a new client connection to the WebSocket %s, err: %v", target, err)
```

**Review by woshilapin — APPROVED** [review]
So we do it open-source?

Apart from that, nice touch about providing documentation and `config.yml.example`, really good for onboarding.

File: README.md [hunk 8]
```diff
@@ -2,3 +2,69 @@
...
+If you compile from source just create `config.yml` in the root directory, copy the content of [config.yml.example](./config.yml.example) and change the variables.
+
+### Configuration parameters
+
+| Environment variable       | parameter in yaml file | Default          | Signification                         |
+| -------------------------- | ---------------------- | ---------------- | ------------------------------------- |
+| APP_WEBSTOMP_TARGET        | webstomp.target        |                  | Url for STOMP connection              |
+| APP_WEBSTOMP_LOGIN         | webstomp.login         |                  | Login                                 |
+| APP_WEBSTOMP_PASSCODE      | webstomp.passcode      |                  | Password                              |
+| APP_WEBSTOMP_DESTINATION   | webstomp.destination   |                  | Topic or queue                        |
```
[the remark this answers is not in the document]
Comment (datanel) [reply README.md:48]: I've separated webstomp client and rabbitmq client and I used the proper terms for them.
If we have a struct that mixes webstomp and rabbitmq, yes we may have something like "source". 
but if we expose the lib I don"t think "source" is the good choice

File: config.yml.example [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+  passcode: "your_passcode"
+  destination: "/topic/a_topic"
+  protocol: "v12.stomp"
+  sendTimeout: "20000"
+  recvTimeout: "20000"
+rabbitmq:
+  url: "amqp://guest:guest@localhost:5672//"
+  contentType: "application/json"
+  exchange:
+    name: "your_exhange_name"
```
[the remark this answers is not in the document]
Comment (datanel) [reply config.yml.example]: i always make the mistake

File: internal/rabbitmq/rabbitmq.go [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+		logError("connection to rabbitmq failed, retrying in 1 sec...", err)
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func (m *AmqpManager) reconnector() {
+	for {
+		err := <-m.errorChannel
+		if !m.closed {
+			logError("reconnnecting after connection closed", err)
```
[the remark this answers is not in the document]
Comment (datanel) [reply internal/rabbitmq/rabbitmq.go]: fixed

File: internal/webstomp/webstomp.go (hunk 7 above)
Comment (datanel) [reply internal/webstomp/webstomp.go]: fixed

**Review by pascal-vs — APPROVED** [review]

**Review by pbougue — APPROVED** [review]
Look good.
Although I don't trust myself on golang at all for now.

**Pull request merged**